### PR TITLE
AE: cleanup the mess in AESinkFactory

### PR DESCRIPTION
**Pull request by FernetMenta**
see title

@popcornmix @afedchin @peak3d please register your sinks according to the example of OSX

### CODE REVIEW

**Comment by lrusak** [comment]
Looks great!

Are you going to do the registration for alsa and pulse also?

**Comment by FernetMenta** [comment]
I will prepare alsa for registration and register it for x11 and wayland. Mayby pulse too 😁

**Comment by fritsch** [comment]
One needs to make sure that windowing is initialized before AE is started. On Linux that is not the case. Therefore - even when registering the new Factory in X11 Init (which adds specific dependencies to windowing, e.g. Pulse, ALSA) it will time out and later on register.

Is that what we want? If Windowing should be responsible for the sinks - we need to make really sure that Windowing is up and running and has registered our sinks in the factory succesfully. 

Towards headless this feels like a step back.

Find the AESinkPULSE changes here:  https://github.com/fritsch/xbmc/commit/ec15e2903e18d9d9bc75a0db95dd82ecc7fda6e1

For the no good windowing hook, see here - but we need to change that: https://github.com/fritsch/xbmc/commit/c951705b066045d199c79446968dccdc53cc7759

**Comment by FernetMenta** [comment]
@lrusak ALSA and PA are ready to register. Could you please adapt gbm?

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/windowing/gbm/OptionalsReg.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
+/*
+ *      Copyright (C) 2005-2017 Team XBMC
+ *      http://xbmc.org
```
Comment (MartijnKaijser) [inline xbmc/windowing/gbm/OptionalsReg.cpp]: Maybe update the url in both files to all kodi.tv

Comment (FernetMenta) [reply xbmc/windowing/gbm/OptionalsReg.cpp]: done

**Comment by lrusak** [comment]
What else are we waiting for?

**Comment by FernetMenta** [comment]
android and aml

**Comment by FernetMenta** [comment]
Btw: we could factor out the linux optionals into a linux common file

**Comment by FernetMenta** [comment]
I would like to merge this on the weekend regardless of broken platforms. A 10 days grace period for platform devs to adapt is imo good enough. if no dev from the team cares, merging and breaking raises interest amongst devs in the community.

**Comment by notspiff** [comment]
Do it

9. des. 2017 11:51 skrev "Rainer Hochecker" <notifications@github.com>:

**Comment by MartijnKaijser** [comment]
It seems you forget other platform devs also have lives.

**Comment by koying** [comment]
@fritsch maintains the android sink ;)

**Comment by koying** [comment]
On a personal note, I agree with @fritsch that injecting an AE dependency into the windowing code seems a strange idea

**Comment by FernetMenta** [comment]
there is no AE dependecy in windowing, at leat not for the platforms I take care

**Comment by koying** [comment]
Well, https://github.com/xbmc/xbmc/pull/13119/files#diff-0d41b56c6bf02a9153509f45c8cd71d6R29 is what I call injecting a dependency...

**Comment by afedchin** [comment]
windowing has already a `dependency` to VideoPlayer, so I don't see a big problem

**Comment by fritsch** [comment]
@afedchin what has Videoplayer to do with AE? Or was the argument: the mess is already there let's just add upon it :-) ?

As already said to fernet: I don't think this architectural decision is a good one - but I won't stand in its way. I don't have time to come up and test something different. The abstract ifdeffery factory was a nightmare - I think we can agree on that.

Android PR incoming after succesfully runtime tested. I cannot do AML as this might be combined with the Optional code.

**Comment by koying** [comment]
Hehe... Yep. Wasn't pinged on that one, or I would have made the same remark ;)
We even have a dep on RetroPlayer :)

Are we moving the spaghetti from CApplication to windowing, now :P

**Comment by FernetMenta** [comment]
> Well, https://github.com/xbmc/xbmc/pull/13119/files#diff-0d41b56c6bf02a9153509f45c8cd71d6R29 is what I call injecting a dependency...

that is OSX, X11 and Wayland i.e. do it differently. It's the platform's maintainers decision how to handle it.

**Comment by fritsch** [comment]
Android will look the same as OSX: https://github.com/fritsch/xbmc/commit/58b67ef73d3624470699963cc89948c14634e683 - on Android it does not hurt - it's just a matter of taste. Linux got much better with the optionals that encapsulate the dependencies.

**Comment by koying** [comment]
@FernetMenta Maybe I took "register your sinks according to the example of OSX" too literally ;)
Anyway, indirecting the dependency via OptionalsReg.h doesn't remove the dependency...

**Comment by fritsch** [comment]
If one takes the assumption that Windowing == Platform, one can mathematically say that the platform has to care for its audio. And with the assumption that windowing == platform, it is fine.

Btw. on most systems, especially Android, OSX, Windows - windowing is in fact the platform. Here that makes sense. For linux and headless this assumption is a bit different. As said above: It's a matter of taste.

**Comment by FernetMenta** [comment]
As already said, platform devs are free to register their sinks where they like it best. Of course not with an ifdef in generic code. Currently windowing represents the platform and for the time being this is for most platforms the bast location.

**Comment by koying** [comment]
> If one takes the assumption that Windowing == Platform

> Currently windowing represents the platform

So we have both "platform" and "windowing" having the same concept, now. Interesting...

@fritsch Don't you think https://github.com/xbmc/xbmc/tree/master/xbmc/platform/android/activity would be better suited, rather than windowing?

**Comment by fritsch** [comment]
@koying I put it wherever you want it. Yes I think that's better - but hey the class you are point out is also a "Sammelbecken for everything".

**Comment by koying** [comment]
> the class you are point out is also a "Sammelbecken for everything".

Yep, but at least it's meant to be so, as the glue between the android platform and Kodi ;)

**Comment by fritsch** [comment]
@FernetMenta This will kill the AESinkNULL right? We don't use it as fallback anymore?

**Comment by FernetMenta** [comment]
Nullsink is an unmaintained relic. I think there's a comment somewhere saying that nullSink should not be required with ActiveAE.

**Comment by candrews** [comment]
This change causes Kodi to coredump at startup.
5760c75843 exhibits the coredump; 5760c75843~ does not.
This is an Linux/amd64 system, I can provide any additional info you'd like.
Here are the last lines logged before the crash:
```
21:04:48.676 T:140717323974400   DEBUG: Thread AESink start, auto delete: false
21:04:48.676 T:140717332367104   DEBUG: CSettingsManager: requested setting (audiooutput.dspaddonsenabled) was not found.
21:04:48.676 T:140717323974400    INFO: CActiveAESink::OpenSink - initialize sink
21:04:48.676 T:140717323974400   DEBUG: CActiveAESink::OpenSink - trying to open device PULSE:default
21:04:48.676 T:140717323974400    INFO: CAESinkALSA::Initialize - Attempting to open device "default"
21:04:48.676 T:140717323974400    INFO: CAESinkALSA - Unable to open device "default" for playback
21:04:48.676 T:140717323974400   ERROR: CAESinkALSA::Initialize - failed to initialize device "default"
```

Here's the stack trace:
```
Process 8337 (kodi-x11) of user 1001 dumped core.
#6  0x00005574df7ad04c _ZN11CAESinkALSA10InitializeER13AEAudioFormatRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE (kodi-x11)
#7  0x00005574df7adb3d _ZN11CAESinkALSA6CreateERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEER13AEAudioFormat (kodi-x11)
#8  0x00005574df7732fd _ZN2AE14CAESinkFactory6CreateERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEER13AEAudioFormat (kodi-x11)
#9  0x00005574df79ae80 _ZN8ActiveAE13CActiveAESink8OpenSinkEv (kodi-x11)
#10 0x00005574df79bdf5 _ZN8ActiveAE13CActiveAESink12StateMachineEiPN5Actor8ProtocolEPNS1_7MessageE (kodi-x11)
#11 0x00005574df79c41d _ZN8ActiveAE13CActiveAESink7ProcessEv (kodi-x11)
#12 0x00005574def3a7fd _ZN7CThread6ActionEv (kodi-x11)
#13 0x00005574def3aaa8 _ZN7CThread12staticThreadEPv (kodi-x11)
#1  0x00005574def388d5 _ZN11XbmcThreads11CEventGroup4waitEj (kodi-x11)
#2  0x00005574dec5a5ec _ZN14CRemoteControl7ProcessEv.part.9 (kodi-x11)
#3  0x00005574def3a7fd _ZN7CThread6ActionEv (kodi-x11)
#4  0x00005574def3aaa8 _ZN7CThread12staticThreadEPv (kodi-x11)
#1  0x00005574dee72eef _ZN5Actor8Protocol18SendOutMessageSyncEiPPNS_7MessageEiPvi (kodi-x11)
#2  0x00005574df78b62b _ZN8ActiveAE9CActiveAE8InitSinkEv (kodi-x11)
#3  0x00005574df7908d0 _ZN8ActiveAE9CActiveAE9ConfigureEP13AEAudioFormat (kodi-x11)
#4  0x00005574df7932c8 _ZN8ActiveAE9CActiveAE12StateMachineEiPN5Actor8ProtocolEPNS1_7MessageE (kodi-x11)
#5  0x00005574df7945bd _ZN8ActiveAE9CActiveAE7ProcessEv (kodi-x11)
#6  0x00005574def3a7fd _ZN7CThread6ActionEv (kodi-x11)
#7  0x00005574def3aaa8 _ZN7CThread12staticThreadEPv (kodi-x11)
#1  0x00005574df65dd49 _ZN12ANNOUNCEMENT20CAnnouncementManager7ProcessEv (kodi-x11)
#2  0x00005574def3a7fd _ZN7CThread6ActionEv (kodi-x11)
#3  0x00005574def3aaa8 _ZN7CThread12staticThreadEPv (kodi-x11)
#1  0x00005574dec62cf7 _ZN15CFDEventMonitor7ProcessEv (kodi-x11)
#2  0x00005574def3a7fd _ZN7CThread6ActionEv (kodi-x11)
#3  0x00005574def3aaa8 _ZN7CThread12staticThreadEPv (kodi-x11)
#1  0x00005574dee72eef _ZN5Actor8Protocol18SendOutMessageSyncEiPPNS_7MessageEiPvi (kodi-x11)
#2  0x00005574df78aa34 _ZN8ActiveAE9CActiveAE10InitializeEv (kodi-x11)
#3  0x00005574df22adb8 _ZN12CApplication6CreateERK15CAppParamParser (kodi-x11)
#4  0x00005574def74d46 XBMC_Run (kodi-x11)
#5  0x00005574de9f725d main (kodi-x11)
#7  0x00005574dea5f35a _start (kodi-x11)
```

**Comment by fritsch** [comment]
@t6  Mind adjusting SNDIO? Currently it's not used anymore in the new OptionalsReg.cpp which is used to add it to the Factory on start.

**Pull request merged**